### PR TITLE
Live: Keep same query key on refresh

### PR DESCRIPTION
**Pull request by ryantxu**
This PR updates the live measurements code so we use the same key after a refresh, rather than a new one. 

The main effect is that hitting "refresh" will continue to send the same key into the PanelQueryRunner stream, so any existing caches should be maintained

### CODE REVIEW

**Comment by ryantxu** [comment]
My manual testing seems OK -- I suspect we forced a new key every time to force cleanup, but the unnecessary cleanup can lose state from previous values, so it is not great with labeled values

File: public/app/plugins/datasource/grafana/datasource.ts [hunk 1]
```diff
@@ -135,14 +134,14 @@ export class GrafanaDatasource extends DataSourceWithBackend<GrafanaQuery> {
...
-            buffer,
-          })
-        );
+        const opts = {
+          key: `live-${target.refId}-`,
+          addr: addr!,
+          filter,
+          buffer,
+        };
+        opts.key += djb2Hash(JSON.stringify(opts)); // Keep a constant key for the same query+options
```
Comment (leeoniya) [inline public/app/plugins/datasource/grafana/datasource.ts:143]: can we cache the opts string instead of doing `JSON.stringify(opts)` each time?

Comment (ryantxu) [reply public/app/plugins/datasource/grafana/datasource.ts:143]: it is a nested structure, so i guess we can walk it iterating keys+values?  but not sure that is any better

File: public/app/plugins/datasource/grafana/datasource.ts [hunk 2]
```diff
@@ -135,14 +134,14 @@ export class GrafanaDatasource extends DataSourceWithBackend<GrafanaQuery> {
...
-        results.push(
-          getGrafanaLiveSrv().getDataStream({
-            key: `${request.requestId}.${counter++}`,
-            addr: addr!,
-            filter,
-            buffer,
-          })
-        );
+        const opts = {
+          key: `live-${target.refId}-`,
```
Comment (ArturWierzbicki) [inline public/app/plugins/datasource/grafana/datasource.ts:138]: should this also change here? https://github.com/grafana/grafana/blob/963fffeef10cdad8276e46c6a20f2a13b0491c31/public/app/features/live/centrifuge/service.ts#L193

Comment (ArturWierzbicki) [inline public/app/plugins/datasource/grafana/datasource.ts:138]: should we also change it here? https://github.com/grafana/grafana/blob/963fffeef10cdad8276e46c6a20f2a13b0491c31/public/app/features/live/centrifuge/service.ts#L193

Comment (ArturWierzbicki) [reply public/app/plugins/datasource/grafana/datasource.ts:138]: right - but the change in the PR only applies to the Grafana DS, wouldn't we want to apply it to all ds?

Comment (toddtreece) [reply public/app/plugins/datasource/grafana/datasource.ts:138]: are there going to be collisions if two panels have streams with the same ref ID?

**Comment by toddtreece** [comment]
dashboard: [telegraf-test-1707317685421.json](https://github.com/grafana/grafana/files/14195427/telegraf-test-1707317685421.json)

run these in separate terminal sessions:
```
telegraf --config one.conf
```

```
telegraf --config two.conf
```

one.conf:
```
[global_tags]
  app = "one"

[agent]
  interval = "3s"
  flush_interval = "5s"

[[inputs.mem]]

[[inputs.cpu]]
  percpu = false
  totalcpu = true

[[outputs.http]]
  url = "http://localhost:3000/api/live/push/one"
  data_format = "influx"
  [outputs.http.headers]
    Authorization = "Bearer glsa_xxxxxx"
```

two.conf:
```
[global_tags]
  app = "two"

[agent]
  interval = "3s"
  flush_interval = "5s"

[[inputs.mem]]

[[inputs.cpu]]
  percpu = false
  totalcpu = true

[[outputs.http]]
  url = "http://localhost:3000/api/live/push/two"
  data_format = "influx"
  [outputs.http.headers]
    Authorization = "Bearer glsa_xxxxxx"
```

File: public/app/plugins/datasource/grafana/datasource.ts (hunk 2 above)
Comment (ryantxu) [reply public/app/plugins/datasource/grafana/datasource.ts:138]: not here -- each panel has its PanelQueryRunner that is managing the 'key'

@ArturWierzbicki -- i think you are right... we may want a better default for all datasources, but lets make sure it is fixed for the GrafanaDS before expanding scope

**Comment by ryantxu** [comment]
this does not really fix the issue... just makes it less clear when it happens 😢

**Pull request closed without merging**